Fix torch.profiler Schedule Function (Function Event only)

sraikund16 opened the pull request:
Summary:
github issue: https://github.com/pytorch/pytorch/issues/73828

Whenever we transition from RECORD_AND_SAVE to WARMUP in the profiler schedule, we instantiate a new backend profiler which wipes out the last cycle's information. This makes using the `repeat` parameter less useful in the schedule as you only get contents of the last cycle/repeat. In this diff, we save the accumulated Function Events before setting the new ones and then merge the two EventLists after post processing/cleaning is done. This diff only fixes Function Events so that we can get statistics over each cycle within a schedule. A follow up should be made to accumulate the chrome tracings as well if it is requested.


Test Plan: Added functional python tests in test_profiler.py that test different schedules and their FunctionEvent counts

Differential Revision: D56956245




cc @ezyang @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

aaronenyeshi reviewed (commented):
Can you please check failed unit test: profiler/test_profiler.py::TestProfiler::test_schedule_function_count. Looks like the test that was added. 

Also, is there a potential for CPU OOM if the user sets repeat=0, and function_events grow unbounded?

sraikund16 commented:
> Can you please check failed unit test: profiler/test_profiler.py::TestProfiler::test_schedule_function_count. Looks like the test that was added.
> 
> Also, is there a potential for CPU OOM if the user sets repeat=0, and function_events grow unbounded?

Should be fixed now. 

In terms of a CPU OOM, repeat = 0 has the same upper bound as not specifying scheduler, because in that case we only use RECORD. So in this case, it doesn't look like we are exposing any vulnerability that is any different than whatever is available now

aaronenyeshi reviewed (approved):
Sounds good to me!

sraikund16 committed:
Fix torch.profiler Schedule Function (Function Event only) (#125510)

Summary:

github issue: https://github.com/pytorch/pytorch/issues/73828

Whenever we transition from record and save to warmup, we instantiate a new backend profiler which wipes out the last cycle's information. In this diff, we save the accumulated Function Events before setting the new ones and then merge the two EventLists after post processing/cleaning is done. This diff only fixes Function Events so that we can get statistics over each cycle within a schedule. A follow up should be made to accumulate the chrome tracings as well if it is requested.

Test Plan: Added functional python tests in test_profiler.py that test different schedules and their FunctionEvent counts

Reviewed By: aaronenyeshi

Differential Revision: D56956245
